src/lib.rs

```diff
@@ -93,25 +93,64 @@ macro_rules! try_get_param {
 /// The `broker_handler!` macro calls this on error. We don't use a `From`
 /// implementation, because this way we get app and request context, and we
 /// don't necessarily have to pass the error on to Iron.
-fn handle_error(app: &AppConfig, _: &mut Request, err: BrokerError) -> IronResult<Response> {
-    match err {
-        BrokerError::Input(_) => {
-            let obj = ObjectBuilder::new()
-                .insert("error", err.description())
-                .build();
-            let content = serde_json::to_string(&obj).unwrap();
-            let content_type = modifiers::Header(ContentType::json());
-            Ok(Response::with((status::BadRequest, content_type, content)))
-        }
-        BrokerError::Provider(_) => {
-            // TODO: Redirect to RP with the error description
-            Err(IronError::new(err, status::ServiceUnavailable))
+fn handle_error(app: &AppConfig, req: &mut Request, err: BrokerError) -> IronResult<Response> {
+    // Build the complete redirect, if we're far enough into the request to
+    // have have a valid redirect_uri from the RP.
+    let redirect_uri = req.extensions.remove::<RedirectUri>().map(|base| {
+        let mut url = base.into_generic_url();
+        {
+            let mut pairs = url.query_pairs_mut();
+            pairs.append_pair("error", match err {
+                BrokerError::Input(_) => "invalid_request",
+                BrokerError::Provider(_) => "temporarily_unavailable",
+                _ => "server_error",
+            });
+            match err {
+                BrokerError::Input(_) | BrokerError::Provider(_) => {
+                    pairs.append_pair("error_description", err.description());
+                },
+                _ => {}
+            }
         }
-        _ => {
+        url.into_string()
+    });
+    // Based on the kind of error, we may log the error, hide the description
```

Comment:
I think this comment could be expanded a bit. As I understand it, there are two basic resulting scenarios: (a) render the error template with the description set in the previous block (with appropriate status code), (b) redirect to the RP, with a specific URL also set in the previous block. The current comment makes it sound like there are three different things going on, which I find hard to trace/understand?

Also, maybe add a blank line before this whole block?